### PR TITLE
initial order1 #goodnessSquad

### PR DESCRIPTION
**Pull request by amir-arad**
use external d.ts files instead of internals
init react state within lifecycle
fix minor typings errors

### CODE REVIEW

File: scripts/src/core/utils.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ module fugazi.utils {
 	
 		reset(): void {
 			this.cancel();
-			this.name = setTimeout(this.callback, this.interval);
+			this.name = setTimeout(this.callback, this.interval) as any as number;
```
Comment (nitzantomer) [inline scripts/src/core/utils.ts]: why is this needed?

File: scripts/src/view/input/base.tsx [hunk 2]
```diff
@@ -55,13 +55,16 @@ namespace fugazi.view.input {
...
 
-			this.state.value = props.value || "";
 
 			this.prompt = prompt;
 			this.className = ["input", className];
 			this.keymap = collections.map<() => void>();
 		}
 
+		public componentWillMount() {
+			this.setState({value : this.props.value || ""});
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx]: why do it here instead of the ctor?

File: scripts/src/view/input/base.tsx [hunk 3]
```diff
@@ -250,12 +253,16 @@ namespace fugazi.view.input {
 		public constructor(props: P, className: string, prompt?: string) {
 			super(props, "suggestible " + className, prompt);
 
-			this.state.showing = false;
-			this.state.message = null;
-			this.state.suggestions = this.props.suggestions;
-			this.state.focus = "input";
 		}
 
+		public componentWillMount() {
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx]: again, why not in the ctor?

File: scripts/src/view/input/base.tsx [hunk 4]
```diff
@@ -328,11 +335,11 @@ namespace fugazi.view.input {
 		private element: HTMLElement;
 		private isFocused: boolean;
 
-		constructor(props: SuggestionPanelProperties) {
-			super(props);
-
+		public componentWillMount() {
 			this.key = utils.generateId();
-			this.state.selected = -1;
+			this.setState({
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx]: same question

File: scripts/src/view/input/base.tsx [hunk 5]
```diff
@@ -441,7 +448,9 @@ namespace fugazi.view.input {
 		public constructor(props: P, className: string, prompt?: string) {
 			super(props, "bgtext " + className, prompt);
 
-			this.state.backgroundText = props.backgroundText || "";
+			this.setState({
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx]: you haven't moved it to `componentWillMount` but you changed it to use `setState`, why?

File: scripts/src/core/utils.ts (hunk 1 above)
Comment (amir-arad) [reply scripts/src/core/utils.ts]: the nodejs typings says that setTimeout returns `Timer`, which is not a number.
This is a fast workaround

File: scripts/src/view/input/base.tsx (hunk 2 above)
Comment (amir-arad) [reply scripts/src/view/input/base.tsx]: the new react typings, with the new typescript capabilities, define `this.state` and all its fields as constant. 
in addition, calling `this.setState(...)` at ctor issues a React console warning about component lifecycle.
This is the correct way (by react) to set an initial state.

File: scripts/src/view/input/base.tsx (hunk 5 above)
Comment (amir-arad) [reply scripts/src/view/input/base.tsx]: my bad, should move to `componentWillMount`

File: scripts/src/view/input/base.tsx [hunk 6]
```diff
@@ -55,11 +56,10 @@ namespace fugazi.view.input {
 		public constructor(props: P, className: string, prompt?: string) {
 			super(props);
 
-			this.state.value = props.value || "";
-
 			this.prompt = prompt;
 			this.className = ["input", className];
 			this.keymap = collections.map<() => void>();
+			this.state = {value: this.props.value || ""} as any;
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx:62]: `as P` instead of `as any` won't do the trick?

File: scripts/src/view/input/base.tsx [hunk 7]
```diff
@@ -249,11 +249,13 @@ namespace fugazi.view.input {
 
 		public constructor(props: P, className: string, prompt?: string) {
 			super(props, "suggestible " + className, prompt);
+			this.state = {
+				showing: false,
+				message: null,
+				suggestions: this.props.suggestions,
+				focus: "input"
+			} as any;
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx:257]: same, `as P`, so we'll have type safety

File: scripts/src/view/input/base.tsx [hunk 8]
```diff
@@ -328,11 +330,13 @@ namespace fugazi.view.input {
...
 
-		constructor(props: SuggestionPanelProperties) {
+		public constructor(props: SuggestionPanelProperties, className: string, prompt?: string) {
 			super(props);
 
 			this.key = utils.generateId();
-			this.state.selected = -1;
+			this.state = {
+				selected: -1
+			} as any;
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx:339]: `as SuggestionPanelProperties`

File: scripts/src/view/input/base.tsx [hunk 9]
```diff
@@ -328,11 +330,13 @@ namespace fugazi.view.input {
 		private element: HTMLElement;
 		private isFocused: boolean;
 
-		constructor(props: SuggestionPanelProperties) {
+		public constructor(props: SuggestionPanelProperties, className: string, prompt?: string) {
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx:333]: what are `className` and `prompt` for?  they are not used

File: scripts/src/view/input/base.tsx [hunk 10]
```diff
@@ -440,13 +444,14 @@ namespace fugazi.view.input {
 	export class BackgroundTextInput<P extends BackgroundTextInputProperties, S extends BackgroundTextInputState> extends InputView<P, S> {
 		public constructor(props: P, className: string, prompt?: string) {
 			super(props, "bgtext " + className, prompt);
-
-			this.state.backgroundText = props.backgroundText || "";
+			this.state = {
+				backgroundText: this.props.backgroundText || ""
+			} as any;
```
Comment (nitzantomer) [inline scripts/src/view/input/base.tsx:449]: `as P`

File: scripts/src/view/main.tsx [hunk 11]
```diff
@@ -1,7 +1,9 @@
-/// <reference path="../../lib/react.d.ts" />
 /// <reference path="../core/utils.ts" />
 /// <reference path="terminal.tsx" />
 
+// import {CSSProperties} from "react";
```
Comment (nitzantomer) [inline scripts/src/view/main.tsx:4]: remove comments

File: scripts/src/view/main.tsx [hunk 12]
```diff
@@ -86,12 +88,12 @@ module fugazi.view {
 				top: 0,
 				right: 0,
 				border: 0
-			}
+			};// as CSSProperties;
```
Comment (nitzantomer) [inline scripts/src/view/main.tsx:91]: remove comment

File: scripts/src/view/renderers.tsx [hunk 13]
```diff
@@ -294,11 +295,15 @@ module fugazi.view.renderers {
 			}
 
 			if (fieldsList && fieldsList.length > 0) {
-				this.state.isStruct = true;
-				this.state.fields = collections.map<components.types.Type>();
+				this.setState({
```
Comment (nitzantomer) [inline scripts/src/view/renderers.tsx:298]: this looks problematic because `setState` isn't synchronous, and three lines below there's this:
`fieldsList.forEach(map => this.state.fields.extend(map));`

File: scripts/src/view/input/base.tsx (hunk 6 above)
Comment (amir-arad) [reply scripts/src/view/input/base.tsx:62]: it's `as S` and it breaks because S is unknown

Comment (nitzantomer) [reply scripts/src/view/input/base.tsx:62]: That's weird, I don't get errors for this.  Maybe due to the new ts version?
Fine, after the PR I'll take this branch and will try to figure it out.
